Analysis/study page keyboard input

brollin opened the pull request:
Let's add the keyboardMove input to the analysis and study page. Not sure that the normal location directly under the board is best, but I implemented it there for the time being:

![image](https://github.com/lichess-org/lila/assets/12655228/4cb0ddef-87aa-4689-a60a-89d8fa229999)

schlawg commented:
I'm really pumped about this. Especially if you're interested in adding voice after! It's been requested several times. Variants may be tricky.

brollin commented:
Totally! Hoping to get voice in there right after. And I also have expectations of bug reports with both given the complexity of analysis+study+variants O_O

schlawg commented on ui/analyse/src/ctrl.ts:
```diff
@@ -418,12 +422,12 @@ export default class AnalyseCtrl {
   }
 
-  mainlinePathToPly(ply: Ply): Tree.Path {
+  mainlinePlyToPath(ply: Ply): Tree.Path {
```
Lol good catch!

brollin replied:
Couldn't unsee it 🙃

brollin commented:
Okay it seems to be wired up correctly, with only one very questionable bit of hackery. As for placement, now that we have something to look at maybe we can collectively decide where it should go. There are some outstanding styling issues that I didn't want to spend to much time on in case we decided on a different location.

brollin commented on ui/analyse/src/ground.ts:
```diff
@@ -6,12 +6,31 @@ import { DrawShape } from 'chessground/draw';
+            data: {
+              ...ctrl.data,
+              // TODO: improve this disgraceful shoehorning
+              player: ctrl.chessground.state.movable as { color: Color },
```
This is the questionable bit. We only use this piece of data (`player.color`) in crazyhouse when checking the pockets for a piece, and up until now it could be a static thing. In analysis because color is a dynamic piece of data, we probably want to just refactor a little bit.

ornicar commented on ui/analyse/src/ground.ts:
```diff
@@ -6,12 +6,27 @@ import { DrawShape } from 'chessground/draw';
+            fen: ctrl.node.fen,
+            canMove: true,
+            cg: ctrl.chessground,
+          });
```
ctrl sub-ctrls should be instanciated within the ctrl, not in a view.

brollin replied:
That definitely feels better. I ended up mimicking how we do things in ui/round: https://github.com/lichess-org/lila/pull/15019/commits/fcbf24f59b6619fe64e81584e1fdda71efb82f40

brollin commented on ui/analyse/src/ctrl.ts:
```diff
@@ -947,4 +965,24 @@ export default class AnalyseCtrl {
+    this.sendMove(orig, dest, capture, prom);
+  };
+
+  auxUpdate = (fen: string) => {
```
I believe the intention was "auxiliary update", but agreed even spelled out it could be improved. Its purpose is to fan out updates to the various plugins, and I think it existing at all will make more sense when we implement voiceMove. We could name it something like `updatePlugins`?

schlawg replied:
pluginUpdate and pluginMove sound good.

brollin replied:
In any case, made that naming change here: https://github.com/lichess-org/lila/pull/15019/commits/10b0d242fbdc789a9679f0db86ea091f8858b258

brollin replied:
As for `pluginMove`, something like `handlePluginMove` makes the most sense to me. If I'm in a controller and see that function, I know that we are handling a move that's coming from the plugin. I'm always reminding myself which way the data is flowing, and in theory this could help with that.

schlawg replied:
the actor followed by the verb is a convention shared by rounds, puzzles, and chessground. i.e userMove, apiMove, etc.

but i don't care at all, i'm the one who chose `auxMove` so just shut me in a soundproof closet and pick something - it's fine.

brollin replied:
That's fair, consistency is important. Too light of a preference to break with consistency or shove anyone in a closet I think. Side note, I've loved seeing the refactored controllers (functions->classes) and overall the code is in a great spot it feels like, a nice departure from my work's codebase 😌

schlawg commented on ui/analyse/tsconfig.json:
```diff
@@ -11,6 +11,7 @@
     { "path": "../chess/tsconfig.json" },
     { "path": "../common/tsconfig.json" },
     { "path": "../game/tsconfig.json" },
+    { "path": "../keyboardMove/tsconfig.json" },
```
The tsc dependency mechanism isn't documented anywhere so I'm impressed. I think you're the first to do this.

brollin replied:
I won't pretend to understand it, but I will pretend to be quite good at pattern matching 😎

schlawg replied:
it's actually straightforward. tsc doesn't care about our monorepo dependencies (package.json), so if you want correct behavior when type checking - you have to either compile everything in the correct order (bad) or express dependencies in tsconfig.json (better, but still not great).

brollin commented:
Now the styling issues aren't as egregious, but with 2 or 3 columns in a crazyhouse study chapter with the keyboard move off, there is a little extra margin. Leaving a note for myself more than anything. Lots to keep track of with all of the possible edge cases.

schlawg commented:
I like it. I did notice these areas where the various keyboard event handlers need to be deconflicted:
- up and down arrow dont navigate the fork control when the keyboard move box has focus
- hitting tab when the keymove box is focused puts you into a never never land where all navigation keystrokes are discarded. enter won't focus again, clicking the board won't either. you must shift tab, click back to the keymove, or click the movelist to get out of it.
- hitting tab or escape on keymove should probably focus movelist so that any keyboard controls that conflict with keymove intercepts can still be issued.

brollin commented:
Good catches. I fixed the up and down arrow behavior in 1a4004b.

It turns out that the computer evaluation toggle was the next thing in the tab order, and when it is focused, or any input element for that matter, mousetrap key bindings are not respected. I solved this two ways:

* I changed the tab order in 06b84be such that the keyboard input was after all of the right hand side tools (ceval, move list, controls, etc) in the tab order. Now, the under board is focused after the keyboard input.
* I allowed toggling the computer evaluation via enter in 7f4517e, only when it is focused. Now if that element happens to be focused and enter is pressed, something happens at least. Since it is an input, it will still be swallowing lots of other key presses (like the arrows).

I tried to move the keyboard input such that the next thing in the tab order would be the move list, but the move list is in the middle of `renderTools`, and the keyboard input is rendered alongside those tools. Basically the hierarchy doesn't easily allow that. Open to changing the tab order around more if desired.

brollin commented on ui/keyboardMove/src/main.ts:
```diff
@@ -122,9 +128,19 @@ export function ctrl(root: KeyboardMoveRootCtrl): KeyboardMove {
+        ArrowLeft: -1,
+        ArrowRight: 1,
+      };
+      root.userJumpPlyDelta?.(arrowKeyToPlyDelta[arrowKey]);
```
There's an opportunity here to just remove `root.userJumpPlyDelta` in favor of `root.handleArrowKey`. I was debating whether it would be cleaner or not 🤔

ornicar commented:
It's looking a bit wrong on /analysis 
![258x685_2024-05-01](https://github.com/lichess-org/lila/assets/140370/a0b36384-59ba-4617-bcc3-b4706ae0ab89)

Also I couldn't get it to appear on a game analysis page?

ornicar commented:
yes, it's on lichess.dev now

brollin commented:
Just catching back up, I can look into why it isn't on the analysis page anymore as well as the styling.

How should I log into lichess.dev? Is there a preexisting user I can use or some way to register? Attempting to register doesn't send me an email at the moment.

brollin commented:
Okay I think it's in a better state now for testing

ornicar commented:
`Uncaught DOMException: Node.insertBefore: Child to insert before is not a child of this node`

This error usually means someone changed the dom that snabbdom manages.

ornicar commented:
Deployed to lichess.dev!

ornicar commented:
ready to go?

brollin commented:
I think so! I'll keep my eyes out for any bug reports.

ornicar commented on ui/analyse/css/study/_layout.scss:
```diff
@@ -1,3 +1,8 @@
 }
+
+.analyse .keyboard-move {
+  margin-top: calc($analyse-block-gap * 2);
```
`#{$...}` syntax?
